Source/Core/Core/IOS/Network/NCD/Manage.cpp

```diff
@@ -29,11 +35,51 @@ IPCCommandResult NetNCDManage::IOCtlV(const IOCtlVRequest& request)
   switch (request.request)
   {
   case IOCTLV_NCD_LOCKWIRELESSDRIVER:
+    if (!request.HasNumberOfValidVectors(0, 1))
+      return GetDefaultReply(IPC_EINVAL);
+
+    if (request.io_vectors[0].size < 2 * sizeof(u32))
+      return GetDefaultReply(IPC_EINVAL);
+
+    if (m_context.ipc_fd != 0)
+    {
+      // It is an error to lock the driver again when it is already locked.
+      common_result = IPC_EINVAL;
+    }
+    else
+    {
+      // NCD writes the internal address of the request's file descriptor.
+      // We will just write the value of the file descriptor.
+      // The value will be positive so this will work fine.
+      m_context.ipc_fd = request.fd;
+      Memory::Write_U32(request.fd, request.io_vectors[0].address + 4);
+    }
     break;
 
   case IOCTLV_NCD_UNLOCKWIRELESSDRIVER:
-    // Memory::Read_U32(request.in_vectors.at(0).address);
+  {
+    if (!request.HasNumberOfValidVectors(1, 1))
+      return GetDefaultReply(IPC_EINVAL);
+
+    if (request.in_vectors[0].size < sizeof(u32))
+      return GetDefaultReply(IPC_EINVAL);
+
+    if (request.io_vectors[0].size < sizeof(u32))
+      return GetDefaultReply(IPC_EINVAL);
+
+    const u32 request_handle = Memory::Read_U32(request.in_vectors[0].address);
+    if (m_context.ipc_fd == request_handle)
+    {
+      m_context.ipc_fd = 0;
+      common_result = 0;
+    }
+    else
+    {
+      common_result = -3;
```

Comment:
Do we know what `-3` is?

Reply:
No, it is an unknown error code. I don't remember it being used elsewhere so I'm not sure how to name it...

Reply:
Are the error codes in [this table](https://github.com/devkitPro/libogc/blob/b4ad94705b8d9d68fb51014615582b3b9f9ae836/libogc/network_wii.c#L157-L235) applicable?  If so, `-3` would be `EADDRINUSE`.

Reply:
I don't think they're related, as this module also uses -4 which is the standard IPC_EINVAL. -3 is probably another standard IOS error code.

Reply:
Do you think it's worth adding it to the ReturnCode enum anyway (named something like `IPC_ERROR_3`), or are you fine with this as is?